source/extensions/filters/listener/kmesh_tlv/kmesh_tlv.cc

```diff
@@ -0,0 +1,112 @@
+#include "kmesh_tlv.h"
+
+#include "source/common/network/address_impl.h"
+#include "source/common/network/utility.h"
+#include "source/common/network/filter_state_dst_address.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace KmeshTlv {
+
+Network::FilterStatus KmeshTlvFilter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(trace, "kmesh_tlv: new connection accepted");
+  cb_ = &cb;
+  // Waiting for data.
+  return Network::FilterStatus::StopIteration;
+}
+
+Network::FilterStatus KmeshTlvFilter::onData(Network::ListenerFilterBuffer& buffer) {
+  const ReadOrParseState read_state = parseBuffer(buffer);
+  switch (read_state) {
+  case ReadOrParseState::Error:
+    cb_->socket().ioHandle().close();
+    return Network::FilterStatus::StopIteration;
+  case ReadOrParseState::TryAgainLater:
+    return Network::FilterStatus::StopIteration;
+  case ReadOrParseState::SkipFilter:
+    return Network::FilterStatus::Continue;
+  case ReadOrParseState::Done:
+    return Network::FilterStatus::Continue;
+  }
+  return Network::FilterStatus::Continue;
+}
+
+ReadOrParseState KmeshTlvFilter::parseBuffer(Network::ListenerFilterBuffer& buffer) {
+  ENVOY_LOG(trace, "kmesh tlv listener filter parse buffer");
+
+  auto raw_slice = buffer.rawSlice();
+  const uint8_t* buf = static_cast<const uint8_t*>(raw_slice.mem_);
+
+  while (raw_slice.len_ >= expected_length_) {
+    ENVOY_LOG(trace, "already has {} bytes in the buffer, expected length is {}", raw_slice.len_,
+              expected_length_);
+
+    switch (state_) {
+    case TlvParseState::TypeAndLength:
+      ENVOY_LOG(trace, "tlv parse state is TypeAndLength");
+
+      if (buf[index_] == TLV_TYPE_SERVICE) {
+        ENVOY_LOG(trace, "process TVL_TYPE_SERVICE");
+
+        uint32_t content_len = 0;
+        std::memcpy(&content_len, buf + index_ + 1, TLV_LENGTH_LEN);
+        content_len = ntohl(content_len);
+        ENVOY_LOG(trace, "get tlv length {}", content_len);
+
+        expected_length_ += content_len;
+        content_length_ = content_len;
+        index_ += (TLV_TYPE_LEN + TLV_LENGTH_LEN);
+        state_ = TlvParseState::Content;
+
+      } else if (buf[index_] == TLV_TYPE_ENDING) {
+        ENVOY_LOG(trace, "process TLV_TYPE_ENDING");
+
+        buffer.drain(expected_length_);
+
+        return ReadOrParseState::Done;
+      } else {
+        ENVOY_LOG(error, "invalid tlv type {}", buf[index_]);
+
+        return ReadOrParseState::Error;
+      }
+      break;
+
+    case TlvParseState::Content:
+      ENVOY_LOG(trace, "tlv parse state is Content");
+
+      sockaddr_storage addr;
+      int len;
+
+      addr.ss_family = AF_INET;
+      len = sizeof(struct sockaddr_in);
+      auto in4 = reinterpret_cast<struct sockaddr_in*>(&addr);
+      std::memcpy(&in4->sin_addr, buf + index_, 4);
```

Comment:
should we double check here if input byte content of a service address is also 4+2 bytes length? in case the L4 send a incorrect address.